python/demo/gsoc2022/demo_scattering_boundary_conditions.py

```diff
@@ -0,0 +1,448 @@
+# # Scattering from a wire with scattering boundary conditions
+# This demo is implemented in two files: one for the mesh
+# generation with gmsh, and one for the variational forms
+# and the solver. It illustrates how to:
+#
+# - Use complex quantities in FEniCSx
+# - Setup and solve Maxwell's equations
+# - Add scattering boundary conditions for transparent boundaries
+# - Implement Scattering Boundary Conditions as absorbing boundaries
+# - Calculate absorption, scattering and extinction efficiencies
+#
+# ## Equations and problem definition
+# Let's consider an infinite metallic wire immersed in
+# a background medium (e.g. vacuum or water). Let's now
+# consider the plane cutting the wire perpendicularly to
+# its axis at a generic point. Such plane $\Omega=\Omega_{m}
+# \cup\Omega_{b}$ is formed by the cross-section
+# of the wire $\Omega_m$ and the background medium
+# $\Omega_{b}$ surrounding the wire. Let's consider
+# just the portion of this plane delimited by an external
+# circular boundary $\partial \Omega$. We want to calculate
+# the electric field $\mathbf{E}_s$ scattered by the wire
+# when a background wave $\mathbf{E}_b$ impinges on it.
+# We will consider a background plane wave at $\lambda_0$
+# wavelength, that can be written analytically as:
+#
+# $$
+# \mathbf{E}_b = \exp(\mathbf{k}\cdot\mathbf{r})\hat{\mathbf{u}}_p
+# $$
+#
+# with $\mathbf{k} = \frac{2\pi}{\lambda_0}n_b\hat{\mathbf{u}}_k$
+# being the wavevector of the
+# plane wave, pointing along the propagation direction,
+# with $\hat{\mathbf{u}}_p$ being the
+# polarization direction, and with $\mathbf{r}$ being a
+# point on the $\Omega$ domain.
+# We will only consider $\hat{\mathbf{u}}_k$ and $\hat{\mathbf{u}}_p$
+# with components belonging
+# to the $\Omega$ domain and perpendicular to each other,
+# i.e. $\hat{\mathbf{u}}_k \perp \hat{\mathbf{u}}_p$
+# (transversality condition of plane waves).
+# If we call $x$ and $y$ the horizontal
+# and vertical axis in our $\Omega$ domain,
+# and by defining $k_x = n_bk_0\cos\theta$ and
+# $k_y = n_bk_0\sin\theta$, with $\theta$ being the angle
+# defined by the propagation direction $\hat{\mathbf{u}}_k$
+# and the horizontal axis $\hat{\mathbf{u}}_x$,
+# we can write more explicitly:
+#
+# $$
+# \mathbf{E}_b = -\sin\theta e^{j (k_xx+k_yy)}\hat{\mathbf{u}}_x
+# + \cos\theta e^{j (k_xx+k_yy)}\hat{\mathbf{u}}_y
+# $$
+#
+# The Maxwell's equation for scattering problems takes the following form:
+#
+# $$
+# \textrm{on } \Omega: \quad-\nabla \times
+# \nabla \times \mathbf{E}_s+\varepsilon_{r} k_{0}^{2} \mathbf{E}_s
+# +k_{0}^{2}\left(\varepsilon_{r}-\varepsilon_{b}\right)
+# \mathbf{E}_{\mathrm{b}}=0,
+# $$
```

Comment:
$$
\quad-\nabla \times
\nabla \times \mathbf{E}_s+\varepsilon_{r} k_{0}^{2} \mathbf{E}_s
+k_{0}^{2}\left(\varepsilon_{r}-\varepsilon_{b}\right)
\mathbf{E}_{\mathrm{b}}=0 \textrm{ in } \Omega,
$$